frame/nfts/src/features/atomic_swap.rs

```diff
@@ -0,0 +1,152 @@
+		item_id: T::ItemId,
+		desired_collection_id: T::CollectionId,
+		desired_item_id: T::ItemId,
+		maybe_price: Option<ItemPrice<T, I>>,
+		maybe_duration: Option<<T as SystemConfig>::BlockNumber>,
+	) -> DispatchResult {
+		let item = Item::<T, I>::get(&collection_id, &item_id).ok_or(Error::<T, I>::UnknownItem)?;
+		ensure!(item.owner == sender, Error::<T, I>::NoPermission);
+
+		Item::<T, I>::get(&desired_collection_id, &desired_item_id)
+			.ok_or(Error::<T, I>::UnknownItem)?;
+
+		let now = frame_system::Pallet::<T>::block_number();
+		let deadline = maybe_duration.map(|d| d.saturating_add(now));
+
+		PendingSwapOf::<T, I>::insert(
+			&collection_id,
+			&item_id,
+			PendingSwap {
+				desired_collection: desired_collection_id,
+				desired_item: desired_item_id,
+				price: maybe_price,
+				deadline,
+			},
+		);
+
+		Self::deposit_event(Event::SwapCreated {
+			collection: collection_id,
+			item: item_id,
+			desired_collection: desired_collection_id,
+			desired_item: desired_item_id,
+			price: maybe_price,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	pub fn do_cancel_swap(
+		sender: T::AccountId,
+		collection_id: T::CollectionId,
+		item_id: T::ItemId,
+	) -> DispatchResult {
+		let item = Item::<T, I>::get(&collection_id, &item_id).ok_or(Error::<T, I>::UnknownItem)?;
+		let swap = PendingSwapOf::<T, I>::get(&collection_id, &item_id)
+			.ok_or(Error::<T, I>::UnknownSwap)?;
+
+		let is_past_deadline = if let Some(deadline) = swap.deadline {
+			let now = frame_system::Pallet::<T>::block_number();
+			now > deadline
+		} else {
+			false
+		};
+
+		if !is_past_deadline {
+			ensure!(item.owner == sender, Error::<T, I>::NoPermission);
+		}
+
+		PendingSwapOf::<T, I>::remove(&collection_id, &item_id);
+
+		Self::deposit_event(Event::SwapCancelled {
+			collection: collection_id,
+			item: item_id,
+			desired_collection: swap.desired_collection,
+			desired_item: swap.desired_item,
+			price: swap.price,
+			deadline: swap.deadline,
+		});
+
+		Ok(())
+	}
+
+	pub fn do_claim_swap(
+		sender: T::AccountId,
+		send_collection_id: T::CollectionId,
+		send_item_id: T::ItemId,
+		receive_collection_id: T::CollectionId,
+		receive_item_id: T::ItemId,
+		maybe_receive_amount: Option<ItemPrice<T, I>>,
+	) -> DispatchResult {
+		let send_item = Item::<T, I>::get(&send_collection_id, &send_item_id)
+			.ok_or(Error::<T, I>::UnknownItem)?;
+		let receive_item = Item::<T, I>::get(&receive_collection_id, &receive_item_id)
+			.ok_or(Error::<T, I>::UnknownItem)?;
+		let swap = PendingSwapOf::<T, I>::get(&receive_collection_id, &receive_item_id)
+			.ok_or(Error::<T, I>::UnknownSwap)?;
+
+		ensure!(send_item.owner == sender, Error::<T, I>::NoPermission);
+		ensure!(swap.desired_collection == send_collection_id, Error::<T, I>::UnknownSwap);
+		ensure!(swap.desired_item == send_item_id, Error::<T, I>::UnknownSwap);
+		ensure!(swap.price == maybe_receive_amount, Error::<T, I>::UnknownSwap);
+
+		if let Some(deadline) = swap.deadline {
```

Comment:
I think a deadline could actually be mandatory for ever swap, with respect to a `MaxDeadlineDuration`.  
This would ensure that no ancient swaps pollute the chain, since we don't take a storage deposit.

Reply:
The should also be means of cleaning the expired swaps up.

Reply:
From my experience, when you make the deadline mandatory, every UI starts putting the max possible value here...
@ruseinov : yep, you can remove any expired swap offer

Reply:
I'd suggest that a year seems not unreasonable. If it's not gone by then then you should have to relist it. We can always extend the max allowed swap length without upsetting people.

Reply:
Done